sympy/integrals/integrals.py

```diff
@@ -444,7 +445,15 @@ def doit(self, **hints):
                     function = factored_function
                 continue
 
-            if function.has(Min, Max):
+            if function.has(Abs, sign) and (
+                (len(xab) < 3 and all(x.is_real for x in xab)) or
+                (len(xab) == 3 and all(x.is_real and x.is_finite for
```

Comment:
Is it necessary to also declare the variable real?

Reply:
For indefinite integrals, yes, this is checked. From reading #2868 it seems there was no consensus on how to deal with `integrate(Abs(x), x)` when x is not declared real, so I left the issue aside.  

In a definite integral the variable is a "dummy variable" so checking assumptions on it should be unnecessary. For example, the behavior of `integrate(1/x, (x, 1, 2))`  is to integrate over the interval [1, 2], not over an arbitrary path in complex plane that goes from 1 to 2, possibly winding around the origin and contributing `2*pi*I*n` to the integral. 

My personal point of view is that indefinite integrals of expressions with Abs or sign do not make really make sense for complex variable x. One could (a) raise an exception; (b) display a warning to the user; (c) return the integral un-evaluated; (d) infer that the domain is real from the form of the function. 

The current behavior is (c). For comparison, `integrate(Max(x, 1), x)` now follows (d) but of course Max(x, 1) is practically an explicit statement that x is real, putting `Abs(x)` in an integral is less so.

Reply:
I agree that indefinite integrals involving abs or sign only make sense for real variables as they are essentially antiderivatives. I was thinking that in a definite integral it might suffice to check the limits.

Reply:
And that's what I do:  `len(xab) == 3 and all(x.is_real and x.is_finite for x in xab[1:])`. Here `xab[1:]` means checking the limits, not the variable (which is `xab[0]`)

Reply:
By the way, I notice test error from assertion that `integrate(cos(k*x)*Abs(x), (x, -pi, pi))` cannot be evaluated. Now it's evaluated (correctly) so the test fails. Funny.

Reply:
Ah, I carelessly interpreted `all(x.is_real and x.is_finite for x in xab[1:])` as saying that *all* components of `xab` should be real, but you are right, of course.